Use ffi_closure_alloc() on Apple silicon to fix crash

Watson1978 opened the pull request:
Fix #800

To prevent crashes, this patch will use `ffi_closure_alloc()` instead of `mmap()` when it run on the Apple silicon platform.
This patch provides another `rbffi_Closure_Alloc` for apple silicon (I'm afraid to affect to other platforms).


This patch needs #801 to succeed the `rake spec` on apple silicon.

Related to https://github.com/CocoaPods/CocoaPods/issues/9907

Watson1978 commented:
Hmm, it is still crashing another place...

```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00000001d6b9f5d0 __pthread_kill + 8
1   libsystem_pthread.dylib       	0x00000001d6c4ac54 pthread_kill + 292
2   libsystem_c.dylib             	0x00000001d6b12624 abort + 104
3   ruby                          	0x00000001024ed048 die + 12 (error.c:647)
4   ruby                          	0x00000001024ed224 rb_bug_for_fatal_signal + 476 (error.c:679)
5   ruby                          	0x0000000102603a50 sigsegv + 96 (signal.c:957)
6   libsystem_platform.dylib      	0x00000001d6c3fc48 _sigtramp + 60
7   libsystem_c.dylib             	0x00000001d6aef09c strdup + 28
8   libcurl.4.dylib               	0x00000001d43f0e08 Curl_setstropt + 76
9   libcurl.4.dylib               	0x00000001d43f2ffc curl_easy_setopt + 8604
10  libffi.dylib                  	0x00000001d44ec050 ffi_call_SYSV + 80
11  libffi.dylib                  	0x00000001d44f49cc ffi_call_int + 944
12  ffi_c.bundle                  	0x0000000102b1979c rbffi_CallFunction + 260 (Call.c:406)
13  ffi_c.bundle                  	0x0000000102b1d3dc attached_method_invoke + 44 (MethodHandle.c:178)
14  libffi.dylib                  	0x00000001d44f4e04 ffi_closure_SYSV_inner + 800
15  libffi.dylib                  	0x00000001d44ec1e8 ffi_closure_SYSV + 56
16  ruby                          	0x000000010268d0f8 vm_call_cfunc_with_frame + 232 (vm_insnhelper.c:2601)
17  ruby                          	0x000000010268d670 vm_call_symbol + 540 (vm_insnhelper.c:2813)
18  ruby                          	0x0000000102670224 vm_sendish + 244 (vm_insnhelper.c:4191) [inlined]
19  ruby                          	0x0000000102670224 vm_exec_core + 11972 (insns.def:799)
20  ruby                          	0x00000001026824c0 rb_vm_exec + 2348
```

larskanis commented:
> Hmm, it is still crashing at another place...

I fear that's caused by the use of `ffi_closure_alloc()`. As mentioned before, FFI made use of it from ffi-1.9.22 to 1.9.24. There have been several issues on MacOS with these versions: https://github.com/ffi/ffi/issues/626 , https://github.com/ffi/ffi/issues/619 , https://github.com/ffi/ffi/issues/634

Also I wonder about `Thread Crashed:: Dispatch queue` which I remember from a similar `fork()` related issue on MacOS: https://github.com/ged/ruby-pg/issues/311#issue-532214355 . The `crash_report.txt` mentions the dispatch queue as well. But I'm not familiar with MacOS at all, so maybe this is completely unrelated.

I thought I added some `fork()` related specs, but noticed, that I forgot about this branch. It is now merged on master as #810. Could you try it?

Watson1978 commented:
> I thought I added some fork() related specs, but noticed, that I forgot about this branch. It is now merged on master as #810. Could you try it?

I rebased this PR with master branch and it works well for me.

<img width="1179" src="https://user-images.githubusercontent.com/199156/87844542-f10fa380-c8f8-11ea-9865-c73525aae84f.png">

Watson1978 commented:
> Hmm, it is still crashing at another place...

I met the crash when I was running [ethon](https://rubygems.org/gems/ethon) gem which depends on this gem.

richdex14 commented:
Hmm I have tried your branch (with both patches applied) and I get a crash even on `rake` or `rake test`. The crash appears to be from here: 
```
/usr/local/lib/ruby/gems/2.7.0/gems/rspec-core-3.9.2/lib/rspec/core/example.rb:257:in `instance_exec'
/Users/dexterrichard/code/tools/ffi/spec/ffi/callback_spec.rb:855:in `block (2 levels) in <module:CallbackInteropSpecs>'
/Users/dexterrichard/code/tools/ffi/spec/ffi/callback_spec.rb:823:in `assert_callback_in_same_thread_called_once'
/Users/dexterrichard/code/tools/ffi/spec/ffi/callback_spec.rb:856:in `block (3 levels) in <module:CallbackInteropSpecs>'
/usr/local/Cellar/ruby/2.7.1_2/lib/ruby/2.7.0/fiddle/import.rb:317:in `bind_function'
/usr/local/Cellar/ruby/2.7.1_2/lib/ruby/2.7.0/fiddle/import.rb:317:in `new'
/usr/local/Cellar/ruby/2.7.1_2/lib/ruby/2.7.0/fiddle/import.rb:317:in `initialize'

```
Any thoughts?

Watson1978 commented:
@richdex14 The Ruby bundled fiddle gem cannot be built properly with Ruby 2.7.

At least when I built Ruby 2.7, I had to add the flags in `configure` command as following.

```
CFLAGS='-g -O0 -DUSE_FFI_CLOSURE_ALLOC=1' ./configure
```

If you install it throgh the Homebrew, I'm thinking it's not working properly.


So, please try with HEAD of https://github.com/ruby/ruby

richdex14 commented:
Ok yes that worked! Built/installed Ruby from HEAD and passes all the tests. The Homebrew version was too old to include that patch, apparently....thanks!

richdex14 commented:
Unfortunately, I am also getting the error with Ethon you described on the earlier post @Watson1978. I am not a super experience Ruby developer (I come from the embedded/hardware world) but let me know if there's anything I can do to help. My end goal is to use cocoa pods fwiw, and cocoapods calls `/usr/local/lib/ruby/gems/2.8.0/gems/ethon-0.12.0/lib/ethon/curls/infos.rb:127:in `easy_getinfo'` triggering the crash

Watson1978 commented:
@richdex14 Unfortunately, `ethon` gem's test now fails as you indicated. The HEAD of Ruby breaks something...

```
$ ruby -v
ruby 2.8.0dev (2020-07-23T14:44:25Z master 098e8c2873) [arm64-darwin20]

$ rake
#<Thread:0x000000013d0255d8 /Users/watson/prj/ethon/spec/support/localhost_server.rb:75 run> terminated with exception (report_on_exception is true):
/Users/watson/.rbenv/versions/2.8.0/lib/ruby/gems/2.8.0/gems/rack-2.2.3/lib/rack/handler/webrick.rb:26:in `run': wrong number of arguments (given 2, expected 1) (ArgumentError)
	from /Users/watson/prj/ethon/spec/support/localhost_server.rb:46:in `boot'
	from /Users/watson/prj/ethon/spec/support/localhost_server.rb:29:in `block in initialize'
	from /Users/watson/prj/ethon/spec/support/localhost_server.rb:75:in `block in concurrently'
/Users/watson/prj/ethon/spec/support/localhost_server.rb:89: warning: constant ::TimeoutError is deprecated

An error occurred in a `before(:suite)` hook.
Failure/Error: raise TimeoutError.new(error_message) if (Time.now - start_time) > timeout

Timeout::Error:
  Boot failed.
# ./spec/support/localhost_server.rb:89:in `wait_until'
# ./spec/support/localhost_server.rb:30:in `initialize'
# ./spec/spec_helper.rb:25:in `new'
# ./spec/spec_helper.rb:25:in `block (2 levels) in <top (required)>'

Finished in 10.43 seconds (files took 1.63 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```

Watson1978 commented:
Seems that [libffi](https://github.com/libffi/libffi) also requires some patches for Apple Silicon (https://github.com/libffi/libffi/pull/577).
((Unfortunately, my issue wouldn't be solved by it))

Watson1978 commented:
I received another reply from support in Apple, like

> You should use ffi_closure_alloc, as you say, it's not just mapping RWX memory on arm64 mac os. codeloc will actually point to a file-backed mapping of /usr/lib/libffi-trampolines.dylib. There's only RX and RW memory involved, no RWX.

in https://github.com/ffi/ffi/issues/800#issuecomment-702895681

felipecsl commented:
@Watson1978 I've tried this change applied on top of latest master of ffi/ffi and seems like it finally worked for me! I was finally able to get a rails project that uses sassc (and thus ffi) to start and work properly :)

antondalgren commented:
> @Watson1978 I've tried this change applied on top of latest master of ffi/ffi and seems like it finally worked for me! I was finally able to get a rails project that uses sassc (which depends on ffi) to start and work properly :)
> For reference, this is what I'm using [felipecsl@17dbdfc](https://github.com/felipecsl/ffi/commit/17dbdfc43d1f6db1cbd2ff14635ffa1a620380a6)

I also did what @felipecsl did with a successful result! 🚀

japodacaj commented:
@Watson1978 This worked for me as well!

larskanis commented on ext/ffi_c/ClosurePool.h:
```diff
@@ -36,6 +36,8 @@ struct Closure_ {
     void* info;      /* opaque handle for storing closure-instance specific data */
     void* function;  /* closure-instance specific function, called by custom trampoline */
     void* code;      /* The native trampoline code location */
+    void* pcl;
```
What does the term `pcl` mean?

Watson1978 replied:
The address for ffi_<p>rep_<c>losure_<l>oc.
Feel free to rewrite it :)

larskanis reviewed (approved):
Thank you @Watson1978, that looks good! I still wonder why the current closure pool doesn't work on arm64-darwin and why it will not be fixed. Anyway it sounds like a good idea to have an alternative mechanism available.

I will add proper switches into `extconf.rb` so that it's possible to force the use of one or another mechanism per `gem install` parameters. The default will be `USE_FFI_ALLOC=1` for arm64-darwin only.

Watson1978 commented:
Just rebased with master branch

larskanis commented:
I merged this manually in commit https://github.com/ffi/ffi/commit/9029cc0da03175c12b412e14f1fdf9814a2cf293 . Thank you @Watson1978 for patience!
